### ISSUE TITLE
[Proposal] add provenance document creation support based on in-toto attestations

### ISSUE THREAD

**Issue by developer-guy** [goreleaser/goreleaser#2737]
**Is your feature request related to a problem? Please describe.**
This is a feature, not related to a problem.

**Describe the solution you'd like**

Attestations are more like a document/record to give details about how the software was built. [in-toto](https://github.com/in-toto/attestation) provides/defines very detailed documents for it, so, GoReleaser can generate software provenance documents based on in-toto attestations IMHO.

[Tekton Chains](https://github.com/tektoncd/chains) is also creating this document based on its [format](https://github.com/tektoncd/chains/blob/main/PROVENANCE_SPEC.md).

**Additional context**

* https://shibumi.dev/posts/introduction-to-in-toto/
* https://blog.acolyer.org/2019/10/02/in-toto/
* https://github.com/in-toto/in-toto-golang

cc: @dlorenc @caarlos0 @dentrax @adityasaky @santiagotorres

**Comment by developer-guy** [comment]
maybe we can approach this issue from a different angle, we can use [slsa-provenance](https://github.com/philips-labs/slsa-provenance-action/blob/main/cmd/slsa-provenance/main.go) CLI from philips-labs to generate an SLSA provenance, then we can sign and attach it using cosign under the hood instead of generating it our own in GoReleaser.

cc: @marcofranssen @shibumi @dentrax

**Comment by caarlos0** [comment]
I think we can experiment running the action before running goreleaser, then just adding the file to the release and signing it

**Comment by caarlos0** [comment]
scratch that, it needs to run after... so, not sure if we use the action directly or integrate it more properly with goreleaser

maybe its possible to use the signs config for it? 🤔

**Comment by SantiagoTorres** [comment]
I think the slsa-provenance type could work (it's still an in-toto type). Let me know if  there is anything I can help w/ it

**Comment by marcofranssen** [comment]
Happy to have a chat on how we can integrate slsa-provenance with goreleaser.

We are using goreleaser as well for releasing the slsa-provenance-action. "chicken and egg" problem to release our own thing with goreleaser, once the provenance feature is added.

**Comment by caarlos0** [comment]
hey @marcofranssen , feel free to ping me on our discord! https://discord.gg/RGEBtg8vQ6

**Comment by caarlos0** [comment]
is there someone interested in working on this?

**Comment by developer-guy** [comment]
There is also another project that can help us. It generates an attestation (based on [SLSA](https://slsa.dev)) and stores it in a Rekor public transparency log server provided by the sigstore community.

＞ https://github.com/slsa-framework/slsa-github-generator-go

**Comment by developer-guy** [comment]
@Dentrax, let's give it a try?

**Comment by developer-guy** [comment]
ping @Dentrax

**Comment by developer-guy** [comment]
helpful links:

* https://github.com/marcofranssen/slsa-workflow-examples
* https://marcofranssen.nl/secure-your-software-supply-chain-using-sigstore-and-github-actions

**Comment by caarlos0** [comment]
@developer-guy lgtm!

**Comment by developer-guy** [comment]
I thought that GoReleaser could use these SLSA provenance generators that were out there, such as [slsa-framework/slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator/blob/675472104f54a100560e96e8090921390a77cadf/.github/workflows/builder_go_slsa3.yml#L84-L88), [philips-labs/slsa-provenance-action](https://github.com/philips-labs/slsa-provenance-action), but @asraa gave another opportunity to make GoReleaser is a part of its provenance generator for building binaries, that is so interesting, and we should discuss on that @caarlos0

**Comment by caarlos0** [comment]
yes, tbh I'm not well versed in all that and don't know how most of it works, so maybe we could chat about it on discord or something (or here too, wherever you prefer 🙏)

**Comment by laurentsimon** [comment]
Please keep me in the loop as well. I work with @asraa on slsa-framework/slsa-github-generator.

**Comment by caarlos0** [comment]
@laurentsimon will probably keep posting here and/or in our [discord](https://discord.gg/RGEBtg8vQ6)

**Comment by Dentrax** [comment]
Has anybody started work on this yet? Maybe we can integrate the slsa-github-generator as a lib.

**Comment by caarlos0** [comment]
I don't think anyone is working on it, no, wanna give it a try?

**Comment by asraa** [comment]
> Has anybody started work on this yet? Maybe we can integrate the slsa-github-generator as a lib.

Hey! Recently @laurentsimon and I have been working on creating an "action wrapper" that reaches SLSA 3, and we thought of GoReleaser Action as the perfect candidate for that.

We are targetting the next quarter or two for GA of it -- it may make sense @Dentrax to integrate with this. If you're interested, we can share how to do that! (It's a lot more lightweight than reconstructing libraries or coding. Just workflow templating)

**Comment by laurentsimon** [comment]
We can showcase a demo in a few weeks. What's the best way to get in touch? slack?

**Comment by caarlos0** [comment]
> We can showcase a demo in a few weeks. What's the best way to get in touch? slack?

I'm on discord, I think @Dentrax is too.

https://goreleaser.com/discord

**Comment by developer-guy** [comment]
count me in !

**Comment by asraa** [comment]
hey! we are one step closer. I have a "POC" that performs a goreleaser build and creates Sigstore signed SLSA 3 provenance.

There's obvious missing parts that I didn't do yet:
* Piping all the possible goreleaser args through the workflow (it's hardcoded with `release`)
* Accessing the default github token to actually create the release
* Attestation to all the subjects produced (I only did the first)
* Call goreleaser action locally instead of pinned at a version [here](https://github.com/asraa/slsa-on-github/blob/320e6bfde07d7566be4e9641879b4dae469afb77/actions/goreleaser-slsa3/action.yml#L42)

Here's the test run: https://github.com/asraa/slsa-on-github-test/actions/runs/4017434808 producing the goreleaser artifacts in `goreleaser` and the attestations in `slsa-attestations-*`. I'm cleaning this up as we go. 

Here is the SLSA 3 goreleaser workflow (that wraps the action): https://github.com/asraa/slsa-on-github/blob/main/.github/workflows/goreleaser_slsa3.yml

Let's chat! I'll definitely need some help writing docs, testing, making examples.

**Comment by laurentsimon** [comment]
Here's another example for anchor sbom-action (the reusable workflow are just copied / pasted):
https://github.com/laurentsimon/sbom-action/blob/feat/slsa-trw/.github/workflows/slsa3.yml
https://github.com/laurentsimon/sbom-action/blob/feat/slsa-trw/internal/sbom-wrapper/action.yml

You can define arguments inputs (mirroring the original Action): 
https://github.com/laurentsimon/sbom-action/blob/feat/slsa-trw/.github/workflows/slsa3.yml#L11-L113

and pass it to the original Action:
https://github.com/laurentsimon/sbom-action/blob/feat/slsa-trw/internal/sbom-wrapper/action.yml#L36-L49

**Comment by caarlos0** [comment]
gently pint: what's the state of this one?

**Comment by laurentsimon** [comment]
We have done an integration for jreleaser, see https://slsa.dev/blog/2023/08/bring-your-own-builder-github and https://github.com/jreleaser/release-action/tree/v1.1.0-java.
We're happy to help you integrate, unless you prefer keeping only this option https://goreleaser.com/blog/slsa-generation-for-your-artifacts/
/cc @aalmiray

**Comment by caarlos0** [comment]
hey @laurentsimon


Yeah, seems like jreleaser's slsa integration is easier to use than the example from that blog post... I think it would be nice to do something similar...


cc/ @developer-guy wdyt?

**Comment by laurentsimon** [comment]
Note that are are also security considerations:
1. Using BYOB, we can technically record all dependencies used in the attestation, reaching L4 level. We don't support it yet but it's a feature we can add in the future
2. The blog post https://goreleaser.com/blog/slsa-generation-for-your-artifacts/ uses the generic generator, which is only able to attest to the workflow+inputs used by the build, but not the exact runner. On the contrary, BYOB (like our own [Go builder](https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/go/README.md
3. )) can attest to the runner to avoid situations like https://johnstawinski.com/2024/01/11/playing-with-fire-how-we-executed-a-critical-supply-chain-attack-on-pytorch/

**Comment by developer-guy** [comment]
great resources, thanks for sharing!

**Comment by laurentsimon** [comment]
np. To be fair on the runner, we have https://github.com/slsa-framework/slsa-github-generator/issues/1868 which may fix it :)

**Comment by laurentsimon** [comment]
Will any of you attend KubeCon Paris?

**Comment by laurentsimon** [comment]
ouch, I did not see the thumb up :/ Anyone attending OSS@NA?

**Comment by adityasaky** [comment]
I'll be there, as will some other in-toto/attestation folks.

**Comment by caarlos0** [comment]
I think we can now simply do this: https://github.com/goreleaser/goreleaser/pull/5474
For docker images, I'm proposing the following: https://github.com/actions/attest-build-provenance/issues/454
